### DISCUSSION TITLE
Docker 部署的一些说明

### DISCUSSION THREAD

**Discussion by VergilGao** [Mrs4s/go-cqhttp#1896] [General]
在 1.0.0-rc4 版本中，docker的部署方式出现了一些变化，因为没有相关的文档（我没写TAT），所以在这里给想要进行docker部署的同学写一个简单的帮助文档，如果有什么问题也可以在这里提出来，方便未来整理一份QA文档。

### 创建用户和用户组
推荐云服务器用户创建一个专门用来跑 go-cqhttp 的用户和组，如果你是个人nas用户，可以直接跳过这一步
```
groupadd -g 1024 qqbot
```
其中 `-g` 后面的1024可以设定一个你自己喜欢的值，`qqbot`也可以改成一个你喜欢的名字
然后创建一个用户
```
useradd -M -g qqbot -s /sbin/nologin -u 1024 qq123456789
```
其中 `-g qqbot`就是你上面创建的用户组名，`-u` 和 `qq123456879` 同理
此时输入 `id qq123456789` 看一下你新建的用户
```
uid=1024(qq123456789) gid=1024(qqbot) groups=1024(qqbot)
```
## 部署docker镜像
输入：
```
docker run -it --name qqbot -v /opt/go-cqhttp:/data -p 5700:5700 -e UID=1024 -e GID=1024 ghcr.io/mrs4s/go-cqhttp:1.0.0-rc4
```
其中 `-v /opt/go-cqhttp:/data` 指的是你把宿主机的 `/opt/go-cqhttp` 目录挂载为容器的 `/data` 目录，所以你可以指定一个你喜欢的宿主机目录
`-e UID=1024 -e GID=1024` 指的是你运行 go-cqhttp 的用户和用户组，如果你是个人nas用户，可以参考其他docker镜像的设置，特别说明一下，极空间nas以root用户来管理全部文件和运行docker程序，我个人并不推荐使用root用户来跑本程序。另外，如果你是unraid用户，可以不用设置这两个环境变量，因为他们的默认值就是 99:100 ，当然设置一下也没关系

## 首次运行
运行之后，根据提示登陆你的QQ号，输入`ctrl-p`后再输入`ctrl-q`就可以将退出交互界面，现在你的docker程序就在后台运行了

## 从 1.0.0-rc3 和更低版本升级
如果你在运行低版本时没有使用 `--user` 参数，那么你的配置文件目录应该是属于 root 用户的，此时你可以按照**创建用户和用户组**这一章节创建好新的用户和用户组，然后使用 `chown` 来将配置文件目录的归属修改为你创建的新用户。再重新部署新版本的镜像即可。

**Comment by Sam5440** [comment]
有些管理平台不支持输入-u参数的情况下应该如何操作呢？能否默认使用root权限防止无法输入的时候出现一些权限错误问题

**Reply by VergilGao** [reply]
我没太明白你的意思，如果你说的是不让创建用户，那你可以看一下虚拟机平台给你分配的用户的uid和gid是多少，运行docker时传入就好了。
如果你说的是运行 docker时的 -u 参数，目前程序运行时传入UID和GID是以环境变量的方式传入的，输入的是 -e 参数。

**Comment by wdvxdr1123** [comment]
可以同步更新到 [iskong/go-cqhttp-docs](https://github.com/ishkong/go-cqhttp-docs) @ishkong @VergilGao

**Reply by VergilGao** [reply]
> 可以同步更新到 [iskong/go-cqhttp-docs](https://github.com/ishkong/go-cqhttp-docs) @ishkong @VergilGao

最近工作比较忙，等我闲下来时整理一下。
